### ISSUE TITLE
Feature request: Allow knit with parameters: browse to a folder

### ISSUE THREAD

**Issue by ramongallego** [rstudio/rmarkdown#1770]
---

By filing an issue to this repo, I promise that

- [x] I have fully read the issue guide at https://yihui.org/issue/.
- [x] I have provided the necessary information about my issue.
    - If I'm asking a question, I have already asked it on Stack Overflow or RStudio Community, waited for at least 24 hours, and included a link to my question there.
    - If I'm filing a bug report, I have included a minimal, self-contained, and reproducible example, and have also included `xfun::session_info('rmarkdown')`. I have upgraded all my packages to their latest versions (e.g., R, RStudio, and R packages), and also tried the development version: `remotes::install_github('rstudio/rmarkdown')`.
    - If I have posted the same issue elsewhere, I have also mentioned it in this issue.
- [x ] I have learned the Github Markdown syntax, and formatted my issue correctly.

I understand that my issue may be closed if I don't fulfill my promises.

**Comment by ramongallego** [comment]
I have posted a related issue in the shiny [repo](https://github.com/rstudio/shiny/issues/2756), as the Rmarkdown knitr with parameters uses the shiny input options.

The feature I am requesting is to allow the parameters to allow for a new input type: a folder. And that the interface to select it is a browser window - in a similar fashion as you do to select an input file.

I understand that this implementation might be dependent on the shiny app allowing to select folders on the first place. For shiny-only purposes, I found this workaround in [github](https://github.com/wleepang/shiny-directory-input). It is beyond my abilities to fork and push an implementation of this in the params.R file

**Comment by ramongallego** [comment]
![Dada2-knitt-instructions](https://user-images.githubusercontent.com/15640943/74271298-7bf0d000-4cc1-11ea-8cc8-3fec2e7dd765.gif)

**Comment by cderv** [comment]
I may have misunderstood your issue but did you already try the `file` type paramater ? It will render a to a shiny `fileInput` and offer a window to browse... 

See the documentation there in Rmarkdown book and the associated example: 
https://bookdown.org/yihui/rmarkdown/params-knit.html#the-interactive-user-interface 

For each param, you can pass some attributes among them the type of input and some parameters for the shiny input, in addition to the value. 

If I did not understand well, can you precise which type of shiny input you think is missing in Rmarkdown ? 
Currently, rmarkdown only uses the input widget that exists in `shiny`. 

Thank you.

**Comment by ramongallego** [comment]
Hey Christophe,

Thanks for your response. I already tried using the file type, but it does
not allow to select folders, only files. And you are right, what it's
needed is a new shiny input. Apparently there is a shiny extension
https://github.com/thomasp85/shinyFiles that will allow that - I don't know
if it is possible to implement it in Rmarkdown

Many thanks

**Comment by yihui** [comment]
Unfortunately the user interface is not extensible at the moment, but we probably should make it possible for users to bring their own input widget in the future: https://github.com/rstudio/rmarkdown/blob/75bbbcc82247338f1528ab699a1356e71b2686fa/R/params.R#L188-L212

**Comment by angelovangel** [comment]
> Before that becomes possible, an awkward workaround is to use a `fileInput`, get the file path, and what you need is `dirname(path)`.

I think the workaround is not going to work  - the shiny-selected file is in a temp directory, most often specified in a `$TMPDIR` variable by the OS.

**Comment by yihui** [comment]
@angelovangel Oh I forgot that. You are absolutely correct!

**Comment by angelovangel** [comment]
But still, it is a nice feature to have, I’ve been looking for a solution for quite some time. Use case - generate reports by not so experienced users and a directory is used as a parameter in the report.

**Comment by lbramley** [comment]
Any progress on this? It would be really useful - I've got a use case similar to that of @angelovangel

**Comment by daynefiler** [comment]
This would be excellent! I found this thread by looking for a solution that would allow the report to indicate WHICH file was used as the input. The shiny functionality that copies the file to a temporary directory works great for security, and not-great for reproducibility/transparency. A custom input would solve my problem.

**Comment by AmyMikhail** [comment]
I'm chiming in with another use case for this.  I'm creating an R markdown report which includes a map that is created with a shapefile.  Unfortunately I cannot include the shapefiles in a sub-directory of my R project as they are restricted access, so the user has to find and upload them from a shared drive.  

I'm using `sf::st_read()` to import the shapefile, which takes a file path to the `.shp` file as its main argument, but then it expects all the other associated files (such as `.dbf` etc.) to be in the same folder.  When using R markdown parameters (or shiny) the problem is that when the `.shp` file is selected with a `fileInput`, it will move **only** the `.shp` file to the temp directory (where it is available to the .Rmd or shiny) and not the rest of the shapefile-associated files.

This is why it would be great to essentially record the file path on the user's computer that they are selecting the `.shp` file from.  This  file path could then be fed into `sf::st_read()` as normal and it would be able to read in the shapefile along with the necessary information from the other files in the same folder (which all have the same name but different file extensions).

An example of how to use the `shinyFiles` widget that allows directory selection is given [here](https://stackoverflow.com/questions/53551352/upload-a-user-picked-directory-in-r-shiny). 

If I understand @yihui suggestion, this would allow co-opting of any widgets available in shiny to R markdown, is that correct?  Would it also allow co-opting of widgets that are not in shiny itself but available from other shiny support packages (such as `shinyFiles` in this case)?  If so, this would be really helpful.

**Comment by yihui** [comment]
> If I understand @yihui suggestion, this would allow co-opting of any widgets available in shiny to R markdown, is that correct? Would it also allow co-opting of widgets that are not in shiny itself but available from other shiny support packages (such as `shinyFiles` in this case)? If so, this would be really helpful.

Correct. Someone has to implement it, though. It's probably not too hard:

https://github.com/rstudio/rmarkdown/blob/054d73538ec6af25822ef756faf5f7a700d8d631/R/params.R#L207-L211

If users provide a custom input, we may just use that input instead of throwing an error.
